### PR TITLE
dedicated turning lanes phase2

### LINKED ISSUE #567: State machine for separate turning lanes.
In #537 I implemented lane arrow separator to allocate dedicated turning lanes. The lane distribution is according the the number of turning lanes in each of the directions⬅️ ⬆️ ➡️. A more preference is given to the left and less preference is given to the right. 

Now when segment has two lanes you automatically separate lanes to(right hand traffic): ⬅️ + ⬆️ || ➡️. This is good for quick-setup of timed traffic lights.
I think it would be nice to be able to switch to ⬅️ || ⬆️ + ➡️ after another alt+click on (https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/538#issuecomment-552149363 ). This would be good for a different setup kind of timed traffic light .

Screenshot (30)

because of a bug the LHD is the opposite way around (https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/564#issuecomment-557771623 ).
Screenshot (0)

The state machine can also dictate how the lanes are distributed. should we give preference to left in one state and preference to forward in another state? should this be in options menu? or should we keep things simple?

### PR DESCRIPTION
**Pull request by kianzarrin**
fixes #567
dedicated turning lanes now has a state-machine of 2-states:

- 2 lanes: dedicated far turn | dedicated near turn
- 3+ lanes: favor left then forward | favor forward then left
- car+bus lanes:  mix bus and car lanes  | calculate bus arrows and car arrows separately  (we can change this later in #752)

Notes:
- I moved code to dedicated UTIL class.
- I polished the code to use the same code path for near and far turns reducing code complexity.

Test:
I have tested it on 1 lanes 2 lanes and 3 lanes and 12 lane roads with and without bus lanes with and without left hand drive.

### CODE REVIEW

**Comment by aubergine10** [comment]
> I have tested it on 1 lanes 2 lanes and 3 lanes and **12 lane roads**

image

**Comment by aubergine10** [comment]
On LHT city:

image

On this road, Alt+Click the bottom segment set near-side dedicated turn lane first, and nothing changed on subsequent Alt+Clicks.

In every combination of roads at junction I've tried so far, first Alt+Click (after resetting the lane arrows for that segment) will often have some effect, but subsequent Alt+Click seems to have no effect.

**Comment by aubergine10** [comment]
Road at top shows default arrows, highlighted road at bottom shows Alt+Click result:

image

Subsequent Alt+Click did not change anything.

**Comment by aubergine10** [comment]
Ignore those two comments I added (now deleted) - had built the wrong PR, doh!!

**Comment by aubergine10** [comment]
Asymmetric roads, drive on left. This is the default vanilla state of the roads:

image

After Alt+Clicking each segment:

image

**Comment by kvakvs** [comment]
There are arcs of angles somewhere in our code, which correspond to arrow directions. I suggest someone find them, check whether they match what the game is using, and maybe move out somewhere to lane arrow manager.
Might be just inconsistent angle arcs.

**Comment by aubergine10** [comment]
Looks like this branch is based on 11.1.0 tag - can you rebase to master?

**Comment by kianzarrin** [comment]
@aubergine10 @kvakvs I merged in the latest master version.

File: TLM/TLM/UI/SubTools/LaneArrowTool.cs [hunk 1]
```diff
@@ -8,9 +8,15 @@ namespace TrafficManager.UI.SubTools {
     using TrafficManager.Manager.Impl;
```
Comment (kvakvs) [inline TLM/TLM/UI/SubTools/LaneArrowTool.cs:8]: Please wait until my LaneArrows branch is merged, GUI interaction and keyboard handling is different there

**Comment by aubergine10** [comment]
Note: Left Hand Traffic

<detail><summary>issue 1 - wrong arrows</summary>
Vanilla:
image

Alt+Click segment in lane connector tool:
image
</detail>

<detail><summary>issue 2 - wrong arrows</summary>
Vanilla:
image

Alt+Click:
image
</detail>

<detail><summary>issue 3 - segment selection</summary>
With this vanilla road, viewed from this camera angle, there was a selection dead-zone (indicated by red dashed area) where the mouse was not detecting the road.

image

If camera was rotated to pretty much any other angle, the road selection worked normally.
</detail>

**Comment by kianzarrin** [comment]
> issue 3 - segment selection
With this vanilla road, viewed from this camera angle, there was a selection dead-zone (indicated by red dashed area) where the mouse was not detecting the road. 

I will check it out but most likely needs a issue with saved game and fixed camera attached to it.
would be interesting to see if NetAdjust panel can hover the segment properly.

**Comment by kianzarrin** [comment]
I fixed the bug in my code. Now lane arrows should be right.

**Comment by kianzarrin** [comment]
> issue 3 - segment selection
With this vanilla road, viewed from this camera angle, there was a selection dead-zone (indicated by red dashed area) where the mouse was not detecting the road.

@aubergine10 I didn't manage to reproduce this. Can you please create city with minimum test case and attach it to issue (please create new issue)

**Comment by aubergine10** [comment]
Testing again now....

If I manage to recreate issue 3 I'll try and create a save with vanilla + TMPE and nothing else and attach to new issue.

**Comment by kianzarrin** [comment]
This is ready for review.

**Comment by aubergine10** [comment]
@kvakvs can this be merged now?

**Comment by kianzarrin** [comment]
fixes #567 dedicated turning lanes now has a state-machine of 2-states:
-     2 lanes: dedicated far turn | dedicated near turn
-     3+ lanes: favor left then forward | favor forward then left
-     car+bus lanes: mix bus and car lanes | calculate bus arrows and car arrows separately
Notes:
- I moved code to dedicated UTIL class.
- I polished the code to use the same code path for near and far turns reducing code complexity.

**Pull request merged**